ddtrace/contrib/aiohttp/patch.py

```diff
@@ -1,39 +1,374 @@
+                span.parent_id = self._self_parent_span_id
+
+            _set_request_tags(span, _get_url_obj(self))
+            result = yield from self.__wrapped__.read(*args, **kwargs)  # noqa: E999
+            span.set_tag(ext_http.STATUS_CODE, self.status)
+            span.set_tag('Length', len(result))
+
+        return result
+
+    if PY_35:
+        @asyncio.coroutine
+        def __aenter__(self):
+            result = yield from self.__wrapped__.__aenter__()  # noqa: E999
+            return result
+
+        @asyncio.coroutine
+        def __aexit__(self, exc_type, exc_val, exc_tb):
+            result = yield from self.__wrapped__.__aexit__(exc_type,  # noqa: E999
+                                                           exc_val, exc_tb)
+            return result
+
 
+class _WrappedRequestContext(wrapt.ObjectProxy):
+    def __init__(self, obj, pin, span, trace_headers, trace_context):
+        super().__init__(obj)
+        pin.onto(self)
+        self._self_span = span
+        self._self_trace_headers = trace_headers
+        self._self_trace_context = trace_context
+        self._self_have_context = False
 
-def patch():
+    @asyncio.coroutine
+    def _handle_response(self, coro):
+        try:
+            resp = yield from coro  # noqa: E999
+
+            if self._self_trace_headers:
+                tags = {hdr: resp.headers[hdr]
+                        for hdr in self._self_trace_headers
+                        if hdr in resp.headers}
+                self._self_span.set_tags(tags)
+
+            self._self_span.set_tag(ext_http.STATUS_CODE, resp.status)
+            self._self_span.error = int(500 <= resp.status)
+            return resp
+        except BaseException:
+            self._self_span.set_traceback()
+            raise
+        finally:
+            if not self._self_have_context or not self._self_trace_context:
+                self._self_span.finish()
+
+    # this will get when called without a context
+    @asyncio.coroutine
+    def __iter__(self):
+        resp = yield from self._handle_response(self.__wrapped__.__iter__())  # noqa: E999
+        return resp
+
+    if PY_35:
+        def __await__(self):
+            resp = yield from self._handle_response(  # noqa: E999
+                self.__wrapped__.__await__())
+            return resp
+
+        @asyncio.coroutine
+        def __aenter__(self):
+            self._self_have_context = True
+            resp = yield from self._handle_response(  # noqa: E999
+                self.__wrapped__.__aenter__())
+            return resp
+
+        @asyncio.coroutine
+        def __aexit__(self, exc_type, exc_val, exc_tb):
+            try:
+                resp = yield from self.__wrapped__.__aexit__(exc_type, exc_val,  # noqa: E999
+                                                             exc_tb)
+            finally:
+                if self._self_have_context and self._self_trace_context:
+                    self._self_span.finish()
```

Comment:
Maybe define a `_finish_span` method that can be used here and in L195 to avoid repeating?